services/ns_net/allocation_builder.cc

```diff
@@ -0,0 +1,77 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/services/ns_net/allocation_builder.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+namespace mojo {
+
+AllocationBuilder::AllocationBuilder()
+    : buffer_(nullptr), buffer_length_(0), data_length_(0) {}
+
+AllocationBuilder::~AllocationBuilder() {
+  free(buffer_);
+}
+
+uint32_t AllocationBuilder::Size() const {
+  return data_length_;
+}
+
+static inline uint32_t RoundPowerOfTwo(uint32_t x) {
+  if (x == 0) {
+    return 1;
+  }
+
+  --x;
+
+  x |= x >> 1;
+  x |= x >> 2;
+  x |= x >> 4;
+  x |= x >> 8;
+  x |= x >> 16;
+
+  return x + 1;
+}
+
+bool AllocationBuilder::Append(const uint8_t* data, uint32_t length) {
```

Comment:
Yeah, as I said in the final paragraph, controlling how we read from the socket is the correct way to go. Something like starting a two phase write to the pipe, checking how much we can fit into it and reading up to those number of bytes from the underlying socket. `NSURLConnection` does not afford us this level of control however. Should we rewrite it using raw socket APIs? We can also then use the same implementation on Android.